File: tauri-tail-maker/src/utils/logger.ts
import { invoke } from '@tauri-apps/api/core'
import { listen } from '@tauri-apps/api/event'
import type { LogEntry, LogConfig, LogLevel } from '../types/logger'

interface AppEventPayload {
  level: string
  target: string
  message: string
  data?: unknown
}

class Logger {
  private initialized = false
  private buffer: LogEntry[] = []
  private flushTimer: ReturnType<typeof setTimeout> | null = null
  private listeners: Array<(entry: LogEntry) => void> = []

  async init(config: LogConfig = {}) {
    if (this.initialized) return

    await invoke('init_logger', { config })

    // 监听统一后端事件
    await listen<AppEventPayload>('app:event', (event) => {
      const { level, target, message } = event.payload
      const entry: LogEntry = {
        level: level as LogLevel,
        message,
        target,
        timestamp: new Date().toISOString(),
      }
      // 回显到浏览器 console
      const fn = (console as any)[level] ?? console.log
      fn(`[${target}] ${message}`)
      // 通知外部监听者
      this.listeners.forEach((cb) => cb(entry))
    })

    this.initialized = true
    this.flushBuffer()
  }

  private flushBuffer() {
    if (this.buffer.length === 0) return
    const entries = [...this.buffer]
    this.buffer = []
    invoke('emit_logs', { entries }).catch(console.error)
  }

  private scheduleFlush() {
    if (this.flushTimer) clearTimeout(this.flushTimer)
    this.flushTimer = setTimeout(() => this.flushBuffer(), 50)
  }

  log(level: LogLevel, message: string, target = 'frontend') {
    const entry: LogEntry = {
      level,
      message,
      target,
      timestamp: new Date().toISOString(),
    }
    this.buffer.push(entry)
    this.scheduleFlush()
  }

  info(message: string, target?: string) { this.log('info', message, target) }
  warn(message: string, target?: string) { this.log('warn', message, target) }
  error(message: string, target?: string) { this.log('error', message, target) }
  debug(message: string, target?: string) { this.log('debug', message, target) }
  trace(message: string, target?: string) { this.log('trace', message, target) }

  onLog(fn: (entry: LogEntry) => void): () => void {
    this.listeners.push(fn)
    return () => {
      const idx = this.listeners.indexOf(fn)
      if (idx !== -1) this.listeners.splice(idx, 1)
    }
  }

  flush() {
    if (this.flushTimer) {
      clearTimeout(this.flushTimer)
      this.flushTimer = null
    }
    this.flushBuffer()
  }
}

export const logger = new Logger()

File: tauri-tail-maker/src-tauri/src/logger.rs
use shared::logger::{self, LogEntry, LogLevel, LogObserver};
use std::sync::Arc;
use tauri::AppHandle;

use crate::events;

// ── 前端观察者 ──────────────────────────────────────────────

/// 把日志通过统一的 `app:event` 推送到前端
struct FrontendLogObserver {
    app: AppHandle,
}

impl LogObserver for FrontendLogObserver {
    fn on_log(&self, entry: &LogEntry) {
        let level = format!("{:?}", entry.level).to_lowercase();
        if let Some(ref data) = entry.data {
            events::emit_data(&self.app, &level, &entry.target, &entry.message, data.clone());
        } else {
            events::emit_log(&self.app, &level, &entry.target, &entry.message);
        }
    }
}

// ── Tauri 命令 ──────────────────────────────────────────────

#[derive(Debug, serde::Deserialize)]
pub struct LogConfig {
    /// 最低日志级别（"trace" | "debug" | "info" | "warn" | "error"），默认 info
    pub min_level: Option<String>,
}

/// 初始化日志系统 —— 前端启动时调用一次
///
/// 配置日志级别 + 注册前端观察者，后续所有后端日志会通过 `log:entry` 事件推送到前端。
#[tauri::command]
pub fn init_logger(app: AppHandle, config: LogConfig) -> Result<(), String> {
    let dispatcher = logger::global_dispatcher().ok_or("Logger not initialized")?;

    // 设置最低级别
    if let Some(ref level_str) = config.min_level {
        let level = parse_level(level_str)?;
        dispatcher.set_min_level(level);
    }

    // 订阅前端观察者
    let observer = Arc::new(FrontendLogObserver { app });
    dispatcher.subscribe(observer);

    Ok(())
}

/// 前端批量推送日志到后端队列
///
/// 前端同步写入的日志通过此命令进入后端异步分发通道。
#[tauri::command]
pub fn emit_logs(entries: Vec<LogEntry>) {
    if let Some(d) = logger::global_dispatcher() {
        for entry in entries {
            d.log(entry);
        }
    }
}

// ── 辅助函数 ────────────────────────────────────────────────

fn parse_level(s: &str) -> Result<LogLevel, String> {
    match s.to_lowercase().as_str() {
        "trace" => Ok(LogLevel::Trace),
        "debug" => Ok(LogLevel::Debug),
        "info" => Ok(LogLevel::Info),
        "warn" => Ok(LogLevel::Warn),
        "error" => Ok(LogLevel::Error),
        other => Err(format!("Invalid log level: {}", other)),
    }
}
